### DISCUSSION TITLE
Ragged data in Awkward vs. Scipp

### DISCUSSION THREAD

**Discussion by SimonHeybrock** [scikit-hep/awkward#1663] [General]
Hi!
I am working on [Scipp](https://scipp.github.io), which supports some types of ragged data (see [Scipp Documentation on Binned Data](https://scipp.github.io/user-guide/binned-data.html) for details).

Given the interest in pydata/xarray#4285 for using Awkward arrays with Xarray, which is conceptually very close to what `scipp.DataArray` can do, I thought it is high time to kick off a discussion between us. I think there are a lot of similarities but also differences and it would be valuable to understand those, which might at some point even lead to support of a common API or functionality. Please also have a look in https://discuss.scientific-python.org/t/ragged-array-summit/465.

For starters here are high level similarities and differences I am aware of (note that I have used Awkward for little more than 10 minutes):

- Scipp's binned data is conceptually very similar to `ak.layout.ListArray64`. We store an ND-array of start and stop indices, and a "buffer" (I think Awkward calls this "content")? I managed to convert scipp binned data into an Awkward list array without copying the buffer/content.
- Scipp does *not* support nesting or similar more complex cases.
- Scipp supports (1-D) `scipp.DataArray` and `scipp.Dataset` (similar to `xarray.DataArray/Dataset`) as the content. This seems similar to using records in Awkward"s content, with slightly more structure. In particular, we distinguish data fields and coordinate fields (and only data fields are operated on in, e.g., arithmetic operations). This does not really matter in terms of data layout and memory handling, i.e., one could imagine swapping out the internals of how things are stored.
- Scipp provides operations for (re)grouping and (re)binning ragged data in multiple dimensions.
- Scipp is multi-threaded out-of-the-box (based on TBB). I could not find info on whether Awkward does this (or whether it relies on and can make use of multi-threading in the underlying (numpy) array library?

**Comment by agoose77** [comment]
> In particular, we distinguish data fields and coordinate fields (and only data fields are operated on in, e.g., arithmetic operations). 

How does this behave in practice? Do you throw errors when attempting arithmetic operations on coordinate fields? 

Awkward does not have this distinction at the type level, instead we allow users to define how records should be added; we recently changed our model so that records by default do not add together, but library authors can opt-in by writing an implementation. You can look at the [vector library](https://vector.readthedocs.io/en/latest/) as an example of this, although it's a bit indirect as a consequence of supporting multiple "backends" besides Awkward. The [design prototype](https://vector.readthedocs.io/en/latest/usage/vector_design_prototype.html) perhaps shows how vector uses Awkward Array in a trivial sense.

I wonder @jpivarski @ioanaif if there are merits for inverting this logic, so that the default case works, and behavior authors can opt-out? I believe one would do this by setting an `np.ufunc` rule for their behaviors. I don't have a strong opinion yet.

> Scipp is multi-threaded out-of-the-box (based on TBB). I could not find info on whether Awkward does this (or whether it relies on and can make use of multi-threading in the underlying (numpy) array library?

Awkward's CPU-kernels are currently single-threaded. We don't use NumPy for many of our "kernels" (the compiled loops that underpin the high-level operations and reductions). We are working towards CUDA support, which would give us local parallelism, although we do not yet have all the requried kernels implemented there ([see our roadmap](https://github.com/scikit-hep/awkward/wiki#gpu-backend)). I'm not totally sure on what kind of work we'll be doing to this in the near future (as in, I genuinely don't know). @jpivarski will have an answer.

Separately, we also have `dask-awkward` being developed, which will give us distributed computation (and, through multiple-processes, core-scaling). Clearly this is not the same as local parallelism, but it does somewhat address the problem.

**Reply by SimonHeybrock** [reply]
> > In particular, we distinguish data fields and coordinate fields (and only data fields are operated on in, e.g., arithmetic operations).
> 
> How does this behave in practice? Do you throw errors when attempting arithmetic operations on coordinate fields?

No, because the field itself does not know that it is a coordinate. This is simply done in the bigger context. Consider this example:

```python
# this example is written for dense data arrays, but the same happens conceptually
# when working with ragged data
da1 += da2  # computes da1.values += da2.values *after* ensuring, e.g., that `da1.coords['x'].values`
            # is identical to `da2.coords['x'].values`
da1.coords['x'] += da2.coords['x']  # works, just operate on coord values
````

> Awkward does not have this distinction at the type level, instead we allow users to define how records should be added;

Can you point me to where/how this can be customized?

> > Scipp is multi-threaded out-of-the-box (based on TBB). I could not find info on whether Awkward does this (or whether it relies on and can make use of multi-threading in the underlying (numpy) array library?
> 
> Awkward's CPU-kernels are currently single-threaded. We don't use NumPy for many of our "kernels" (the compiled loops that underpin the high-level operations and reductions). We are working towards CUDA support, which would give us local parallelism, although we do not yet have all the requried kernels implemented there ([see our roadmap](https://github.com/scikit-hep/awkward/wiki#gpu-backend)). I'm not totally sure on what kind of work we'll be doing to this in the near future (as in, I genuinely don't know). @jpivarski will have an answer.
> 
> Separately, we also have `dask-awkward` being developed, which will give us distributed computation (and, through multiple-processes, core-scaling). Clearly this is not the same as local parallelism, but it does somewhat address the problem.

👍

**Reply by agoose77** [reply]
Right, so this is a structure-level decision, when operating on the dataset (in your terminology) rather than the atoms (fields). 

To overload how records are added together, we use our `ak.behavior` system, which maintains a table mapping NumPy ufuncs to overloads, and additionally contains special metadata (like `__array__`, which maps to a user-defined class for custom views over arrays):

https://awkward-array.readthedocs.io/en/latest/ak.behavior.html#overriding-numpy-ufuncs-and-binary-operators

You can see an actual implementation of this [in the vector source code](https://github.com/scikit-hep/vector/blob/4e756fe2cd8a3430260e1cf4cf008874cfc1bb3c/src/vector/backends/awkward.py#L1510)

**Reply by jpivarski** [reply]
Hi! This is very interesting and I followed up on your call for a ragged array summit.

It looks to me like Awkward Array and Scipp are very different things:

  * My take is that Scipp is in a category of libraries with xarray, AnnData, and Pandas, which aim to streamline data analysis by making common tasks easy: units/metadata, propagation of uncertainties, all of those histogramming/sparse data options, parallel processing, etc.
  * Awkward Array aims to be in a category with NumPy: foundational, sometimes used directly by data analysts, sometimes as a backend by other user-facing libraries. (Although I originally intended Awkward to be user-facing, its first applications in high energy physics were usually mediated by the [Coffea](https://coffeateam.github.io/coffea/) library and current discussions with xarray and AnnData are making it sound like it's more useful for us to provide this low-level support. But I did intend from the beginning for Awkward Array to be "a thing like NumPy.")

Some examples of this are the behavior-overriding mechanism that @agoose77 pointed to, and we're focusing on single-threaded computation so that [dask-awkward](https://github.com/ContinuumIO/dask-awkward/) can parallelize it in the horizontal direction (across many threads/processes/computers in a cluster). For vertical acceleration, @agoose77 pointed to the future GPU work: Awkward Arrays are built on top of rectilinear arrays from other libraries and we can swap NumPy for CuPy or JAX to do GPU computations. In the future, all of our structure-changing kernels (compiled as single-threaded C routines) will be ported to GPU equivalents so that the same operations can be performed on CPU and GPU.

I should point out that the kernels we write ourselves are only the ones involved in structure-changing operations, such as broadcasting an array of lists with an array of lists of lists of lists. For any numeric elementwise calculations, we let NumPy (or CuPy, JAX, ...) calculate it. Any parallel processing or vectorization that the underlying "flat math" library can do, we get for free.

Some of these computation functions can be built using JIT-compilation. Awkward has a Numba interface, so that users can write imperative code to iterate over an array of data structures. Some problems are easier to express imperatively than in an array-oriented way, and these implementations are often faster because they make only one pass over the data. We also have a new interface to C++ JIT-compilation using [ROOT](https://root.cern/)'s RDataFrame, though this is most likely of interest to high energy physicists.

So, we're likely working in different problem spaces, but if you're interested in interoperability, then there might be some interesting projects to collaborate on. At minimum, maybe users would want an easy (and zero-copy) way to convert data between the two libraries. Scipp has a lot of metadata, just as xarray does: if these are both targets for us, it would tell us what we would need to do to preserve that metadata through conversions, generalizing #1391.

**Reply by SimonHeybrock** [reply]
> It looks to me like Awkward Array and Scipp are very different things:
> 
>  * My take is that Scipp is in a category of libraries with xarray, AnnData, and Pandas, which aim to streamline data analysis by making common tasks easy: units/metadata, propagation of uncertainties, all of those histogramming/sparse data options, parallel processing, etc.
> 
>  * Awkward Array aims to be in a category with NumPy: foundational, sometimes used directly by data analysts, sometimes as a backend by other user-facing libraries. (Although I originally intended Awkward to be user-facing, its first applications in high energy physics were usually mediated by the [Coffea](https://coffeateam.github.io/coffea/) library and current discussions with xarray and AnnData are making it sound like it's more useful for us to provide this low-level support. But I did intend from the beginning for Awkward Array to be "a thing like NumPy.")
>
> [...]
> So, we're likely working in different problem spaces, but if you're interested in interoperability, then there might be some interesting projects to collaborate on.

I think you summarized this quite well. But Scipp actually also had to cover parts of the second item: Since there was no "thing like NumPy" [matching our requirements] that could provide ragged arrays we developed this on our own. So this is a current point of duplication (which is not to imply that this is necessarily wrong).

Beyond that, there is the question whether we have functionality in Scipp that would be valuable, e.g., to the community Awkward is serving. Aside from the xarray/DataArray discussion, there are binning/grouping algorithms that Scipp provides. These allow for mapping *table -> array of tables* or *array of tables -> array of tables* (where I mean table as a list of records). Our current implementation is restricted to operating on Scipp data types, but conceptually there is nothing major that prevents a rewrite which could operate on, e.g., Python dicts of Numpy-like buffers.

**Reply by SimonHeybrock** [reply]
> Right, so this is a structure-level decision, when operating on the dataset (in your terminology) rather than the atoms (fields).

Yes. Having looked at the `ak.behavior` system, it looks like Scipp is essentially encoding a behavior for one particular content type (scipp.DataArray).

> To overload how records are added together, we use our `ak.behavior` system, which maintains a table mapping NumPy ufuncs to overloads, and additionally contains special metadata (like `__array__`, which maps to a user-defined class for custom views over arrays):
> 
> https://awkward-array.readthedocs.io/en/latest/ak.behavior.html#overriding-numpy-ufuncs-and-binary-operators

I think it may be quite simple to replicate Scipp's behavior with this system, nice! One question though: I tried creating an Awkward array from "slicable" objects other than `numpy.ndarray`, for example:

```python
ak.unflatten(scipp_dataarray, counts)  # has __getitem__ supporting int and slice
```
but this raises an exception. What are the requirements for being able to do this? Basically, I am trying to see whether I can get this to work with a combination of Scipp and `ak.ListArray`, without having to decompose the Scipp object into a plain Python dict for use with `ak.RecordArray`.

**Reply by agoose77** [reply]
Small disclaimer: we're about (in the next few months+) to move to the next major version of Awkward Array, which moves a lot of logic into the Python layer. Although this shouldn't make too many user-facing changes, there are places where things aren't identical. For now, I'll discuss the v2 API, which lives under the `awkward._v2` submodule. If you use the latest pre-releases (1.10.0rc1 as of writing, but we'll soon move to the 2.0 series), then you can try it out. I'll write out the full-name of API functions to make it easier to find them.

Most of the (v2) high-level (`ak._v2`) API uses `ak._v2.operations.ak_to_layout.to_layout` under the hood to convert arbitrary Python objects into the known `ak._v2.contents.Content` objects. If you don't pass anything that Awkward recognises as an existing array object, e.g. an Awkward `Content`, or a NumPy array, then it falls back upon `ak._v2.operations.ak_from_iter.from_iter` to actually build the Array from an iterable object. So, if the `scipp_dataarray` isn't one of these subclasses, then we'll be falling back onto this pathway. It sounds like these objects aren't iterable, so we're failing to find any means of converting them. Note that, even if `scipp_dataarray` _were_ iterable, we probably wouldn't want `from_iter` to be the mechanism that Awkward uses to convert it into an Awkward Array; `from_iter` has to loop over (in C++) all of the elements of the array, and allocate intermediary buffers along the way.

This makes me think that we should expand the ability for Awkward to convert from other array formats. As well as handling CuPy and jax, we could add another pathway to try and invoke `__array__` (probably by asking-for-permission for `__array__`). This would support third-party libraries that implement this interface. We also discussed adding our own mechanism inspired by NEP18 to allow third-party libraries to share the Awkward API, and this would also extend to allow third-party array types to support Awkward by defining an `__awkward_array__` method #1126 

Anyway, back to your problem. If you want to build an Awkward Array, then the lowest-level array Awkward wants is flat-columnar arrays, i.e. NumPy arrays, and a schema (form) that tells Awkward how these buffers are used to build the Array layout (see `ak._v2.operations.ak_from_buffers.from_buffers`). You can also build your array manually using the `ak._v2.contents.Content` layout classes, which are the structured nodes that define an Awkward Array. I can elaborate on this process if you need, but I thought I'd keep this reply shorter for now.

Fundamentally, Awkward Array *needs* to understand the type of the data that it operates upon. We can't, for example, operate on arbitrary Python values - we ultimately dispatch to NumPy (or the equivalent array library e.g. CuPy) to compute reductions, etc. At the high level, we restrict the leaf-values of an Array to be 
- numeric types (list, int, complex, ...)
- datetime types
- string/bytes types
and, technically, tuples / records.

**Reply by SimonHeybrock** [reply]
> Fundamentally, Awkward Array _needs_ to understand the type of the data that it operates upon. We can't, for example, operate on arbitrary Python values - we ultimately dispatch to NumPy (or the equivalent array library e.g. CuPy) to compute reductions, etc. At the high level, we restrict the leaf-values of an Array to be
> 
>     * numeric types (list, int, complex, ...)
> 
>     * datetime types
> 
>     * string/bytes types
>       and, technically, tuples / records.

I suppose this is my question: Awkward seems to be able to deal with arbitrary arrays records and provides `ak.behavior` for  customizing operations. So what prevents it from being able to deal with other array-like objects (if the user provides custom behaviors)? For example, one could imagine directly using `pandas.DataFrame` as "content" (without representing it as a dict of NumPy arrays). Right now Awkward seems to only accept subclasses of `Content`, so `pandas.DataFrame` cannot by used directly, but why is that? Is there a fundamental reason?

**Reply by agoose77** [reply]
I won't speak for @jpivarski, who might have a different take on things (and ultimately steers where this project is going), so I'll just put my short, incomplete take here.

At a design-level, Awkward Array is built around the concept of columnar arrays. This is both for performance and for ergonomics, as Awkward was originally intended for use in HEP, where (among a great deal of other things) it is important to be able to read subsets of large datasets by column. So, were we to support non-Numpy leaves, we'd want them to be structureless. The most appropriate generalisation on Awkward's existing behaviour would be to support non-Numpy flat array-like objects (with no "fields") at the leaves. I don't think this would be impossible (but I'll wait for @jpivarski to inevitably correct me ;)). However, outside of the `ufunc` mechanism (for which we delegate to NumPy), most structure / reduction operators involve calling a compiled kernel. Many kernels don't care about the underlying content, but the reduction kernels do. Specifically, the reduction kernels generalise on the traditional reducers from NumPy by reducing across sublists, e.g. whilst NumPy can only handle
```python
>>> np.min([1, 2, 3, 4])
1
```
Awkward supports variable-length sublists, roughly:
```python
>>> min([
    [1, 2, 3, 4],
    [5, 6, 7]
])
[1, 5]
```
Reductions along any axis ultimately invoke the compiled reduction kernels with a 2D array of variable-length sublists. So, were we to generalise the NumpyArray content to support this, there would need to be some kind of mechanism to implement and define these kernels. If these implementations were written in Python, Awkward would be a significant factor slower. 

At a technical level, we assume that the internal Awkward content nodes form a closed set, so we don't have any support for subclassing or modifying these content classes.

We are exploring efforts to go the other way - i.e. Awkward Arrays as Pandas extension types: https://github.com/scikit-hep/awkward/wiki#pandas-integration

**Reply by jpivarski** [reply]
I think it would be reasonable for `to_layout` to recognize objects that have an `__array__` method and call that to get a NumPy array, which it can then wrap in the Content subclass named `NumpyArray`. This should be near the end of the list of types `to_layout` checks, so that (for instance) CuPy arrays don't get copied over to the CPU to turn them into NumPy arrays, which is what happens when you call their `__array__` method. We should think about the consequences of this, because there are a lot of array-like types out there for which calling `__array__` is destructive, such as Pint arrays with units: would users rather get an error when the units are stripped away, at least to know that it's happening?

Then, for instance, a `pandas.Series` would get turned into an `np.ndarray` that can be a leaf-node in an Awkward Array's Content tree. But again, that's information-losing: it removes the Series's index. Maybe we should think about that more before just doing it.

> Right now Awkward seems to only accept subclasses of `Content`, so `pandas.DataFrame` cannot by used directly, but why is that? Is there a fundamental reason?

So yes, that's fundamental. Data in an Awkward Array are not managed by Python, and it's so that we can do operations _like_ the reducer @agoose77 described. Also some advanced slices and restructuring operations.

There is a little flexibility in what type of object holds the numerical data, wrapped by the `NumpyArray` class. Right now, that object can be a NumPy `ndarray`, a CuPy `ndarray`, or a JAX DeviceArray (for autodiff). Whenever we add a new backend, we also define a new "`nplike`" that is a shim to make all of these array libraries duck-typable—out of the box, they're maybe 90% compatible. Then we also have to make sure that our own functions that go beyond the rectilinear array libraries, such as @agoose77's reducer example, can be executed on the backends' data. (Hence, we need to duplicate our suite of CPU functions to also work on GPU pointers, which is [not done yet](https://github.com/scikit-hep/awkward/wiki#gpu-backend).) So the set of backends is slowly and laboriously extended; it's not something in which arbitrary objects can be dropped in. If they were arbitrary objects, there would be a lot of slicing operations, reducers, etc. that we wouldn't be able to implement (without doing them in slow Python for loops).

**Reply by SimonHeybrock** [reply]
> We should think about the consequences of this, because there are a lot of array-like types out there for which calling `__array__` is destructive, such as Pint arrays with units: would users rather get an error when the units are stripped away, at least to know that it's happening?

> There is a little flexibility in what type of object holds the numerical data, wrapped by the `NumpyArray` class. Right now, that object can be a NumPy `ndarray`, a CuPy `ndarray`, or a JAX DeviceArray (for autodiff). Whenever we add a new backend, we also define a new "`nplike`" that is a shim to make all of these array libraries duck-typable—out of the box, they're maybe 90% compatible.

This sounds like you are almost there? If you can (1) work on arbitrary NumPy-like buffers, (2) have a method to request a list/dict of columns from the content (similar to `__array__` you mention, but supporting record-like objects such as `pandas.DataFrame` or `scipp.DataArray`, i.e., returning multiple columns/buffers), and (3) have a way of creating new content (a callable provided by the content?)?

Probably a constructor callable is not sufficient for (3), since in binary operations will also need to operate on the non-column content such as the unit in you `pint` example. In Scipp we have a similar situation: All kernels are applied to array elements as well as the array unit.

**Reply by jpivarski** [reply]
We can operate on arbitrary NumPy-like buffers, so it's not hard to write (for instance) an `ak.from_dataframe` function to zero-copy wrap a Pandas DataFrame. (It would put each `Series.values` in a `NumpyArray`, all the columns into a `RecordArray`, and wrap that up as an `ak.Array`, maybe carry some metadata as `parameters`.) But in such a case, we are controlling what we want to wrap and how we wrap it; I'm wavering on the idea of opening up the floodgates and ingesting anything with an `__array__` method.

Similarly for `scipp.DataArray`. We have a lot of functions named like `ak.from_*` and `ak.to_*`. That would put the conversion in Awkward, though by symmetry, it could also go in Scipp. (I don't have a strong opinion about that. We consider the set of from/to functions to be an open-ended set and can be third-party library-specific.) Zero-copy and round-trip fidelity are goals, but for technical reasons, it might be necessary to give up on one or the other.
